### PR TITLE
Numpy backend implementation of rnn + cleanup of rnn tests

### PR DESCRIPTION
**Pull request by andhus**
### Summary
Update of Numpy backend to follow the actual Keras api for `backend.rnn`.

### Related Issues
As per @gabrieldemarmiesse's request in https://github.com/keras-team/keras/pull/11499

### PR Overview
- [x] This PR requires new unit tests [y] (make sure tests are included)
- [x] This PR requires to update the documentation [y] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y]
- [x] This PR changes the current API [n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Thanks a lot for the PR! We'll also be able to add your implementation in the documentation (see: #11507). I'll review tonight I think.

Could you add `{{np_implementation}}` at the end of the `rnn` docstring in the TF backend? This will display the numpy implementation of `rnn` in the docs. See #11507 if you don't see what I mean.

**Comment by gabrieldemarmiesse** [comment]
Also, we currently have a lot of flaky tests on travis. So I'll be able to review, but the merging might be delayed as only @fchollet can merge PRs which have a failing build.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -3052,10 +3054,6 @@ def get_matching_mask(mask_t, ref_tensor_t):
             'maximum_iterations': input_length}
 
         if mask is not None:
-            if not states:
-                raise ValueError('No initial states provided! '
```
Comment (gabrieldemarmiesse) [inline keras/backend/tensorflow_backend.py]: This is a change of behavior, so this should be approved by @fchollet.

**Comment by gabrieldemarmiesse** [comment]
I only had the time to check the numpy `rnn` and the `check_rnn_operation`. Both look good. I'll continue reviewing later on.

File: keras/backend/tensorflow_backend.py (hunk 1 above)
Comment (andhus) [reply keras/backend/tensorflow_backend.py]: Sure, I'd say it is just a leftover from the previous PR, see [this line](https://github.com/keras-team/keras/pull/11499/files?utf8=%E2%9C%93&diff=unified#diff-a18b8c6a1191d6f49303e0d599ca8c37L3081). This requirement is now a bit arbitrary and is not there in the other backends.

Comment (gabrieldemarmiesse) [reply keras/backend/tensorflow_backend.py]: Thanks for giving the context.

File: keras/backend/numpy_backend.py [hunk 2]
```diff
@@ -185,43 +185,49 @@ def bias_add(x, y, data_format):
...
-        else:
-            h_t1 = 0
-
-        o_t = h_t + h_t1
-        if w_o is not None:
-            o_t = np.dot(o_t, w_o)
-        o.append(o_t)
+        if mask.dtype != np.bool:
+            mask = mask.astype(np.bool)
+        if len(mask.shape) != 2:
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: Could something like
```python
if mask.shape != inputs.shape[:2]:
```
work? I may be wrong. This would be more coherent with the error message.

File: keras/backend/numpy_backend.py [hunk 3]
```diff
@@ -185,43 +185,49 @@ def bias_add(x, y, data_format):
...
+            raise ValueError(
+                'mask should have `shape=(samples, time)`, '
+                'got {}'.format(mask.shape))
+
+        def apply_mask(mask_t, x_t, x_tm1):
+            # expand mask to match number of dimensions of tensor to mask
+            for _ in range(len(x_t.shape) - 1):
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: For expanding dimensions, I suggest
```python
while mask_t.ndim < x_t.ndim:
```
(there might be a `-1` or `+1` somewhere which I forgot)
For reader, it's more explicit what the result will be (in this case mask_t.ndim will be equal to x_t.ndim).

File: keras/backend/numpy_backend.py [hunk 4]
```diff
@@ -185,43 +185,49 @@ def bias_add(x, y, data_format):
...
+                mask_t = np.expand_dims(mask_t, axis=-1)
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: Would it be possible to move this loop outside the `apply_mask` function? Or not all shapes are known in advance?

File: keras/backend/numpy_backend.py [hunk 5]
```diff
@@ -185,43 +185,49 @@ def bias_add(x, y, data_format):
...
+    output_tm1 = np.zeros(output_0.shape)
+    for t in time_index:
+        output_t, states_t = step_function(inputs[:, t], states_tm1)
+        if mask is not None:
+            output_t = apply_mask(mask[:, t], output_t, output_tm1)
+            states_t = [apply_mask(mask[:, t], state_t, state_tm1)
+                        for state_t, state_tm1 in zip(states_t, states_tm1)]
+        output.append(output_t)
+        states_tm1 = states_t
+        output_tm1 = output_t
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: What does `tml` stands for? (It might be a stupid question, be prepared).

File: keras/backend/numpy_backend.py [hunk 6]
```diff
@@ -716,13 +722,16 @@ def _remove_blanks(inds, num_classes):
...
+    return np.expand_dims(x, axis=axis)
+
+
 square = np.square
 abs = np.abs
 exp = np.exp
 log = np.log
 round = np.round
 sign = np.sign
-expand_dims = np.expand_dims
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: Let's make this modification in another PR to keep this structured, since this change seems unrelated to the goal of this PR.

File: keras/backend/numpy_backend.py [hunk 7]
```diff
@@ -185,43 +185,49 @@ def bias_add(x, y, data_format):
...
+    output = []
+    output_0, _ = step_function(inputs[:, 0], initial_states + constants)
+    states_tm1 = initial_states
+    output_tm1 = np.zeros(output_0.shape)
+    for t in time_index:
+        output_t, states_t = step_function(inputs[:, t], states_tm1)
+        if mask is not None:
+            output_t = apply_mask(mask[:, t], output_t, output_tm1)
+            states_t = [apply_mask(mask[:, t], state_t, state_tm1)
+                        for state_t, state_tm1 in zip(states_t, states_tm1)]
```
Comment (gabrieldemarmiesse) [inline keras/backend/numpy_backend.py]: I'm not a super big fan of multi-lines lists comprehension. I believe it defeats the purpose of the list comprehension. You can ignore this comment if you want. It's just a personal thing.

File: keras/backend/numpy_backend.py (hunk 4 above)
Comment (andhus) [reply keras/backend/numpy_backend.py]: It is possible (and more efficient) but I don't know if that it is more readable, then we need to create separate masks for input and each of the states (and do a "+1 thing" either when expanding masks or picking out the mask for one timestep). This is a working implementation:
```
def rnn(step_function, inputs, initial_states,
        go_backwards=False, mask=None, constants=None,
        unroll=False, input_length=None):

    if constants is None:
        constants = []

    if input_length is None:
        input_length = inputs.shape[1]
    assert input_length == inputs.shape[1]
    time_index = range(input_length)
    if go_backwards:
        time_index = time_index[::-1]

    output = []
    output_0, _ = step_function(inputs[:, 0], initial_states + constants)

    if mask is not None:
        if mask.dtype != np.bool:
            mask = mask.astype(np.bool)
        if mask.shape != inputs.shape[:2]:
            raise ValueError(
                'mask should have `shape=(samples, time)`, '
                'got {}'.format(mask.shape))

        def expand_mask(mask, x):
            while mask.ndim < x.ndim:
                mask = np.expand_dims(mask, axis=-1)
            return mask
        output_mask = expand_mask(mask, output_0)
        states_masks = [expand_mask(mask, state) for state in initial_states]

    states_tm1 = initial_states
    output_tm1 = np.zeros(output_0.shape)
    for t in time_index:
        output_t, states_t = step_function(inputs[:, t], states_tm1)
        if mask is not None:
            output_t = np.where(output_mask[:, t:t + 1], output_t, output_tm1)
            states_t = [np.where(state_mask[:, t:t + 1], state_t, state_tm1)
                        for state_mask, state_t, state_tm1 in
                        zip(states_masks, states_t, states_tm1)]
        output.append(output_t)
        states_tm1 = states_t
        output_tm1 = output_t

    return output[-1], np.stack(output, axis=1), states_tm1
```
Which do you prefer?

File: keras/backend/numpy_backend.py (hunk 5 above)
Comment (andhus) [reply keras/backend/numpy_backend.py]: :) "t minus one". In the `for` loop everything is accurately _indexed_ by the variable `t`. So think LaTeX style `s_t` and `s_{t-1}` or `s_{t+1}` but we can't write `-`/`+` in the variable names so use `m`/`p` instead. I like this notation but maybe it is not so obvious after all!

Comment (andhus) [reply keras/backend/numpy_backend.py]: should we go for `prev_states` / `previous_states` or `states_t_minus_1` instead?

File: keras/backend/numpy_backend.py (hunk 6 above)
Comment (andhus) [reply keras/backend/numpy_backend.py]: Sure, the only difference is that default args are not the same as in keras api.

First I used it in the `rnn` implementation - but then I figured it was better to stick to explicit numpy functions for the sake of the "numpy implementation as documentation".

File: keras/backend/numpy_backend.py (hunk 7 above)
Comment (andhus) [reply keras/backend/numpy_backend.py]: I fully see your point - but I don't agree for comprehensions of this, still reasonable, size :) Mainly because it is more clear that it is an atomic "map" operation.
Using a really "flat" coding style helps and is what I prefer (but it is not standard in Keras):
```
states_t = [
    apply_mask(mask[:, t], state_t, state_tm1)
    for state_t, state_tm1
    in zip(states_t, states_tm1)
]
```

File: keras/backend/numpy_backend.py (hunk 4 above)
Comment (gabrieldemarmiesse) [reply keras/backend/numpy_backend.py]: I think we'll go with the new version. Let's make the logic in the index loop as simple as possible (this is the hard part).

File: keras/backend/numpy_backend.py (hunk 5 above)
Comment (gabrieldemarmiesse) [reply keras/backend/numpy_backend.py]: I think that the style you use is making a lot of sense. Let's keep it. For the people who don't know about it (like me), a comment somewhere `# tm1 means 't minus one'` would be enough.

File: keras/backend/numpy_backend.py (hunk 7 above)
Comment (gabrieldemarmiesse) [reply keras/backend/numpy_backend.py]: It's not really the size, more that the variables are defined after their usage when it's multiline. Let's keep things like that. It's just style.

**Comment by andhus** [comment]
Great, I'll complete it shortly. (Note to self) just saw that `constants` are not used in the for loop! Will fix and improve test coverage (!?).

**Comment by gabrieldemarmiesse** [comment]
A test for that would be welcome indeed. Thanks!

**Comment by andhus** [comment]
@gabrieldemarmiesse Made remaining changes according to decisions, including a test for `constants`. Should be ready to go.

**Comment by gabrieldemarmiesse** [comment]
Thanks for the update, I'll review tonight.

File: tests/keras/backend/backend_test.py [hunk 8]
```diff
@@ -177,6 +177,69 @@ def check_composed_tensor_operations(first_function_name,
...
+        assert_allclose(last_output_k, last_output_np, atol=1e-05)
+        assert_allclose(output_k, output_np, atol=1e-05)
+        for s_k, s_np in zip(last_states_k, last_states_np):
+            assert_allclose(s_k, s_np, atol=1e-05)
+
+        last_output_s.append(last_output_k)
+        output_s.append(output_k)
+        last_states_s.append(last_states_k)
+
+    # compare unroll True/False
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: I thought about it last time, and I don't think we need those checks, because you already did everything needed in the numpy implementation. KNP.rnn is garanteed to give the same result for `unroll=True` and `unroll=False` because you don't use the variable `unroll` in the implementation. You already compare against KNP.rnn, so you're good to go. This should simplify this function quite a bit.

Do you see what I mean? Maybe I'm not clear.

Comment (andhus) [reply tests/keras/backend/backend_test.py]: It's clear what you mean, and all things considered I agree - if we also move out the execution of KNP.rnn and run it once.

This is mainly legacy from how the tests look now (prior to this PR), where this comparison is done (although a bit obfuscated and dependent on the order of `kwargs_list`) even though KNP.rnn is independent or unroll.

These things get unclear with the "reference implementation" approach to testing, connecting back to our discussion [here](https://github.com/keras-team/keras/pull/11499#discussion_r230938541). Are we testing _both_ the `KNP` and `K` implementation in this function? Or only `K`,  taking `KNP` as ground truth? If it is the former then it makes sense to run KNP for both `unroll in [True, False]` and assert equality - we shouldn't assume that the argument is not used, rather make sure to expose if that changes. We basically want the latter, but then it could be a problem that there are no tests for KNP as of my previous argumentation. But enough of that for now - I'll fix 👍

**Comment by andhus** [comment]
Thanks, pushed simplification according to "final comment".

**Comment by andhus** [comment]
@gabrieldemarmiesse failing tests are not due to this PR, it should be good to go.

File: tests/keras/backend/backend_test.py (hunk 8 above)
Comment (gabrieldemarmiesse) [reply tests/keras/backend/backend_test.py]: I think it's fine to test for KNP. As you said in our previous discussions, we can make certain assumptions regarding the numpy backend when the function is simple (K.expand_dims for example). In this case, I believe it is simple to check that `unroll` is not used in KNP.rnn because
* `unroll` can have no meaning in the numpy implementation of RNNs.
* With a decent text editor with python pluging / IDE, the variable `unroll` is marked as `unused variable` so it's fairly easy to notice an error.

To be clear:
* I believe that the numpy backend should get tested for complex functions
* KNP.rnn should get tested for correctness
* It is safe to assume that KNP.rnn doesn't use `unroll`

**Comment by gabrieldemarmiesse** [comment]
Sure, we have some issues with flaky tests on travis. See #11660 and #11521 . Github prevents me to merge PRs which have a failing build. Only @fchollet can do it. So either we wait for @fchollet to merge your PR, either we wait for the two PRs which are fixing the flaky tests to be merged. 

Sorry about that, it's really slowing down progress on this repo for everyone.

**Comment by gabrieldemarmiesse** [comment]
Thank you for the PR! It's very helpful!
